### ISSUE TITLE
Events used by multiple modules

### ISSUE THREAD

**Issue by gmoger-jisc** [jiscdev/analytics-udd#352]
How can we manage where an event is used by multiple modules?  Currently we can only assign an event to a single module but these events can actually be shared and are stored as such in timetable systems.

First thoughts on this might be to:
- Remove MOD_INSTANCE_ID from Event
- Student Event should include the MOD_INSTANCE_ID the student is coming from
- New Event relationship table to find which modules and courses it may link to

There may be a more sensible way for this so putting it out for discussion.

**Comment by willblenkhornjisc** [comment]
I think moving MOD_INSTANCE_ID from event to student_event should be enough. Then events are not tied to a MOD_INSTANCE_ID but the student events are.

**Comment by gmoger-jisc** [comment]
If we have a use case to show all events against a module then it would be inefficient to reference the student event table to capture unique event/module relationships.

**Comment by gmoger-jisc** [comment]
I have been looking at this again today and think MOD_INSTANCE_ID and COURSE_INSTANCE_ID will work with the student event entity rather that create a new linking table, as per @willblenkhornjisc suggestion.  I need to explore the API options to see how we might optimise the retrieval of unique events per module without having to pull down all of the data and process on the application/client side.  To be discussed with @ryansmith94

**Comment by ryansmith94** [comment]
I guess you paginate the student events filtering by the MOD_INSTANCE_ID

**Comment by ds10** [comment]
Action from call: 
- Linking table not necessary
- MOD_INSTANCE_ID and COURSE_INSTANCE_ID removed from event 
- MOD_INSTANCE_ID and COURSE_INSTANCE_ID will be added to student event

**Comment by ryansmith94** [comment]
@gmoger-jisc here are the steps I've used for acceptance testing this implementation. Behind the scenes this is using the `group by` in SQL rather than the `select distinct` which we originally planned to use, so please note the change in API for the `distinct` URL parameter which now has the same format as the `select` URL parameter.

### Step 1 - Insert Event

```http
POST /models/event HTTP/1.1
Host: https://jiscudd.learninglocker.net
Authorization: Basic YOUR_BASIC_AUTH
Content-Type: application/json

{
	"EVENT_ID": "EVENT_1",
	"EVENT_NAME": "Example Event"
}
```

### Step 2 - Insert First Student Event

```http
POST /models/studentevent HTTP/1.1
Host: https://jiscudd.learninglocker.net
Authorization: Basic YOUR_BASIC_AUTH
Content-Type: application/json

{
	"EVENT_ID": "EVENT_1",
	"STUDENT_ID": "STUDENT_1",
	"COURSE_INSTANCE_ID": "COURSE_1",
	"MOD_INSTANCE_ID": "MOD_1"
}
```

### Step 3 - Insert Second Student Event

```http
POST /models/studentevent HTTP/1.1
Host: https://jiscudd.learninglocker.net
Authorization: Basic YOUR_BASIC_AUTH
Content-Type: application/json

{
	"EVENT_ID": "EVENT_1",
	"STUDENT_ID": "STUDENT_2",
	"COURSE_INSTANCE_ID": "COURSE_1",
	"MOD_INSTANCE_ID": "MOD_1"
}
```

### Step 4 - Insert Third Student Event

```http
POST /models/studentevent HTTP/1.1
Host: https://jiscudd.learninglocker.net
Authorization: Basic YOUR_BASIC_AUTH
Content-Type: application/json

{
	"EVENT_ID": "EVENT_1",
	"STUDENT_ID": "STUDENT_3",
	"COURSE_INSTANCE_ID": "COURSE_1",
	"MOD_INSTANCE_ID": "MOD_2"
}
```

### Step 5 - Retrieve Distinct Student Events

#### Request
```http
GET /models/studentevent?query={"MOD_INSTANCE_ID": "MOD_1"}&distinct=EVENT_ID&select={"EVENT_ID":true, "event.EVENT_NAME": true}&populate=event HTTP/1.1
Host: https://jiscudd.learninglocker.net
Authorization: Basic YOUR_BASIC_AUTH
```

#### Response
```http
200 OK
Content-Type: application/json; charset=utf-8
x-total-count: 1

[
    {
        "EVENT_ID": "EVENT_1",
        "STUDENT_EVENT_ID": "26d5f521-4155-4df1-a071-5521b0176157",
        "event": {
            "EVENT_NAME": "Example Event"
        }
    }
]
```

**Comment by alanepaull** [comment]
4 Feb 20 call: to be implemented for v1.5.2 dot release this week.
